### PR TITLE
Porting rqt_gui_cpp Cmake

### PR DESCRIPTION
**Pull request by mlautman**
This pr ports rqt_gui_cpp to be built with Colcon 

I intentionally separated this from porting the plugin providers so that we can get a jump start on the review process. Also, seeing as this is my first cpp port, I wanted to get a second pair of eyes on just the cmake changes

### CODE REVIEW

File: rqt_gui_cpp/CMakeLists.txt [hunk 1]
```diff
@@ -23,22 +43,27 @@ set(rqt_gui_cpp_HDRS
...
 
-install(DIRECTORY include/${PROJECT_NAME}/
-  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+install(
+  DIRECTORY include/
+  DESTINATION include
 )
+
+install(FILES plugin.xml
+  DESTINATION share)
```
Comment (brawner) [inline rqt_gui_cpp/CMakeLists.txt]: Package specific destination
```suggestion
  DESTINATION share/${PACKAGE_NAME})
```

File: rqt_gui_cpp/package.xml [hunk 2]
```diff
@@ -12,19 +12,18 @@
...
-  <run_depend version_gte="0.3.0">qt_gui</run_depend>
-  <run_depend version_gte="0.3.0">qt_gui_cpp</run_depend>
-  <run_depend>roscpp</run_depend>
-  <run_depend>nodelet</run_depend>
+  <exec_depend version_gte="0.3.0">qt_gui</exec_depend>
+  <exec_depend version_gte="0.3.0">qt_gui_cpp</exec_depend>
 
   <export>
+    <build_type>ament_cmake</build_type>
     <cpp cflags="-I${prefix}/include" lflags="-L${prefix}/lib -Wl,-rpath,${prefix}/lib -lrqt_gui_cpp"/>
```
Comment (brawner) [inline rqt_gui_cpp/package.xml]: This has some linux specific (or at least non Windows conforming flags). Is there a way to move it to the CMakeLists.txt?

File: rqt_gui_cpp/CMakeLists.txt [hunk 3]
```diff
@@ -1,20 +1,40 @@
...
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CMAKE_AUTOMOC ON)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
+find_package(ament_cmake REQUIRED)
+find_package(Qt5Widgets CONFIG REQUIRED)
+find_package(qt_gui REQUIRED)
+find_package(qt_gui_cpp REQUIRED)
+find_package(rclcpp REQUIRED)
+find_package(tinyxml_vendor REQUIRED)
```
Comment (brawner) [inline rqt_gui_cpp/CMakeLists.txt]: I'm having trouble getting TinyXml to work on Windows. Can we migrate to TinyXml2? It's only a TinyBit outdated

File: rqt_gui_cpp/CMakeLists.txt [hunk 4]
```diff
@@ -23,22 +43,27 @@ set(rqt_gui_cpp_HDRS
 
 qt5_wrap_cpp(rqt_gui_cpp_MOCS ${rqt_gui_cpp_HDRS})
 
-add_library(${PROJECT_NAME} ${rqt_gui_cpp_SRCS} ${rqt_gui_cpp_MOCS})
-target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES} Qt5::Widgets)
+ament_export_dependencies(Qt5Widgets)
 
-if(APPLE)
```
Comment (brawner) [inline rqt_gui_cpp/CMakeLists.txt]: Why remove this?

File: rqt_gui_cpp/CMakeLists.txt (hunk 3 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt]: lol. Would that be a big change? Can you point me to an example migration from TinyXml -> TinyXml2

Comment (brawner) [reply rqt_gui_cpp/CMakeLists.txt]: It depends how extensively it is used, but probably not since only the plugin.xml and the package.xml have to be parsed. 

Check out the ros_pluginlib_plugin_provider.h here.
https://github.com/ros-visualization/qt_gui_core/compare/crystal-devel...ros2_port_windows?expand=1

File: rqt_gui_cpp/package.xml (hunk 2 above)
Comment (mlautman) [reply rqt_gui_cpp/package.xml]: I'm sure there is but I'm not entirely sure why this is here or what the CMakeLists.txt equivalent would look like. Can you weigh in here @dirk-thomas?

File: rqt_gui_cpp/CMakeLists.txt (hunk 3 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt]: It doesn't look like tinyxml is even used in this package. I think I can delete it

File: rqt_gui_cpp/CMakeLists.txt (hunk 4 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt]: restored

File: rqt_gui_cpp/CMakeLists.txt [hunk 5]
```diff
@@ -1,20 +1,37 @@
...
+
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  add_compile_options(-Wall -Wextra -Wpedantic)
+endif()
+
+# Qt5 boilerplate options from http://doc.qt.io/qt-5/cmake-manual.html
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CMAKE_AUTOMOC ON)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
```
Comment (dirk-thomas) [inline rqt_gui_cpp/CMakeLists.txt]: Please use `add_compile_options` instead. I don't think it is supported on all platforms?

File: rqt_gui_cpp/CMakeLists.txt [hunk 6]
```diff
@@ -23,22 +40,34 @@ set(rqt_gui_cpp_HDRS
...
+ament_export_include_directories(include)
+
+add_library(${PROJECT_NAME}
+  ${rqt_gui_cpp_SRCS}
+  ${rqt_gui_cpp_MOCS})
+
+target_link_libraries(
+  ${PROJECT_NAME}
+  ${qt_gui_cpp_LIBRARIES}
+  ${ament_LIBRARIES}
```
Comment (dirk-thomas) [inline rqt_gui_cpp/CMakeLists.txt]: This isn't a valid variable afaik.

File: rqt_gui_cpp/CMakeLists.txt [hunk 7]
```diff
@@ -23,22 +40,34 @@ set(rqt_gui_cpp_HDRS
...
+  Qt5::Widgets
 )
 
-install(TARGETS ${PROJECT_NAME}
```
Comment (dirk-thomas) [inline rqt_gui_cpp/CMakeLists.txt:37]: The library doesn't seem to be installed anywhere anymore?

File: rqt_gui_cpp/package.xml (hunk 2 above)
Comment (dirk-thomas) [reply rqt_gui_cpp/package.xml]: I don't think this information is used anywhere. I can probably be removed.

File: rqt_gui_cpp/include/rqt_gui_cpp/plugin.h [hunk 8]
```diff
@@ -37,18 +37,16 @@
...
 
 /**
  * The base class for C++ plugins which use the ROS client library.
- * A plugin must not call ros::init() as this is performed once by the framework.
- * The name of the ROS node consists of the prefix "rqt_gui_cpp_node_" and the process id.
+ * A plugin must not call rclcpp::init() as this is performed once by the framework.
  */
 class Plugin
   : public qt_gui_cpp::Plugin
-  , public nodelet::Nodelet
```
Comment (dirk-thomas) [inline rqt_gui_cpp/include/rqt_gui_cpp/plugin.h:51]: With the removal of the base class what is the replacement? A node handle provided through the context?

File: rqt_gui_cpp/CMakeLists.txt (hunk 6 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt]: I got this from https://github.com/ros-visualization/qt_gui_core/blob/0c2aaaaa4340f2a810abd91f6f328236c6beb1f9/qt_gui_cpp/src/qt_gui_cpp_shiboken/CMakeLists.txt#L58

@brawner

File: rqt_gui_cpp/CMakeLists.txt (hunk 5 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt]: This is from https://github.com/ros-visualization/qt_gui_core/blob/def6b8d4e28d42523bcd872729d3d2720dc821fb/qt_gui_cpp/src/qt_gui_cpp/CMakeLists.txt#L5

@brawner Did you put this here and do we need to change it to use `add_compile_options`?

File: rqt_gui_cpp/include/rqt_gui_cpp/plugin.h (hunk 8 above)
Comment (mlautman) [reply rqt_gui_cpp/include/rqt_gui_cpp/plugin.h:51]: This PR is only for getting things to compile correctly

File: rqt_gui_cpp/CMakeLists.txt (hunk 6 above)
Comment (brawner) [reply rqt_gui_cpp/CMakeLists.txt]: ${ament_LIBRARIES} is my fault. It's not a valid variable

File: rqt_gui_cpp/CMakeLists.txt (hunk 5 above)
Comment (brawner) [reply rqt_gui_cpp/CMakeLists.txt]: Can't blame me for this one

File: rqt_gui_cpp/CMakeLists.txt (hunk 7 above)
Comment (mlautman) [reply rqt_gui_cpp/CMakeLists.txt:37]: I'm shooting in the dark for a lot of this. I based the new changes off of this. https://github.com/ros2/urdf/blob/ros2/urdf/CMakeLists.txt
Let me know if this is correct or if there is a better example to draw from

File: rqt_gui_cpp/src/rqt_gui_cpp/roscpp_plugin_provider.cpp [hunk 9]
```diff
@@ -36,120 +36,69 @@
...
 namespace rqt_gui_cpp {
 
 RosCppPluginProvider::RosCppPluginProvider()
   : qt_gui_cpp::CompositePluginProvider()
-  , node_initialized_(false)
-  , wait_for_master_dialog_(0)
-  , wait_for_master_thread_(0)
+  , rclcpp_initialized_(false)
 {
+  // if (rclcpp::is_initialized(rclcpp::contexts::get_global_default_context()))
```
Comment (mlautman) [inline rqt_gui_cpp/src/rqt_gui_cpp/roscpp_plugin_provider.cpp]: I need to test this with an up to date ros2 build

File: rqt_gui/src/rqt_gui/ros_plugin_provider.py [hunk 10]
```diff
@@ -64,6 +64,7 @@ def discover(self, discovery_data):
         plugin_descriptors = []
         plugin_file_list = self._find_plugins(self._export_tag, discovery_data)
         for package_name, plugin_xml in plugin_file_list:
+
```
Comment (brawner) [inline rqt_gui/src/rqt_gui/ros_plugin_provider.py:67]: Necessary?

File: rqt_gui/src/rqt_gui/ros_plugin_provider.py [hunk 11]
```diff
@@ -113,7 +114,6 @@ def _find_plugins(self, export_tag, discovery_data):
 
     def _parse_plugin_xml(self, package_name, plugin_xml):
         plugin_descriptors = []
-
```
Comment (brawner) [inline rqt_gui/src/rqt_gui/ros_plugin_provider.py:116]: Necessary?

File: rqt_gui/src/rqt_gui/ros_plugin_provider.py [hunk 12]
```diff
@@ -135,7 +135,6 @@ def _parse_plugin_xml(self, package_name, plugin_xml):
                     'plugin_path': os.path.dirname(plugin_xml),
                     'library_path': library_path,
                 }
-
```
Comment (brawner) [inline rqt_gui/src/rqt_gui/ros_plugin_provider.py:138]: Necessary?

File: rqt_gui_cpp/CMakeLists.txt [hunk 13]
```diff
@@ -23,22 +32,64 @@ set(rqt_gui_cpp_HDRS
...
+  ${Qt5Widgets_INCLUDE_DIRS}
+  ${pluginlib_INCLUDE_DIRS}
+)
+
+add_library(${PROJECT_NAME} SHARED
+  ${rqt_gui_cpp_SRCS}
+  ${rqt_gui_cpp_MOCS})
+
+if(WIN32)
+  target_compile_definitions(${PROJECT_NAME} PRIVATE "URDF_BUILDING_DLL")
```
Comment (brawner) [inline rqt_gui_cpp/CMakeLists.txt]: Where does this come from? Add a comment

File: rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.cpp [hunk 14]
```diff
@@ -34,114 +34,99 @@
...
 
-#include <ros/callback_queue.h>
-#include <ros/ros.h>
-
 #include <stdexcept>
-#include <boost/bind.hpp>
 
 namespace rqt_gui_cpp {
 
 NodeletPluginProvider::NodeletPluginProvider(const QString& export_tag, const QString& base_class_type)
```
Comment (brawner) [inline rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.cpp:41]: Should there be a separate PR to rename these classes?

File: rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.h [hunk 15]
```diff
@@ -86,6 +87,9 @@ class NodeletPluginProvider
     virtual ~RosSpinThread();
     void run();
     bool abort;
+    // Create an executor that will be responsible for execution of callbacks for a set of nodes.
+    // With this version, all callbacks will be called from within this thread (the main one).
+    rclcpp::executors::SingleThreadedExecutor exec_;
```
Comment (brawner) [inline rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.h]: I chose to use a MultiThreadedExecutor on the python equivalent. It might help plugins from not blocking each other.

File: rqt_gui/src/rqt_gui/ros_plugin_provider.py (hunk 10 above)
Comment (mlautman) [reply rqt_gui/src/rqt_gui/ros_plugin_provider.py:67]: visually better

File: rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.cpp (hunk 14 above)
Comment (mlautman) [reply rqt_gui_cpp/src/rqt_gui_cpp/nodelet_plugin_provider.cpp:41]: Yes

**Comment by mlautman** [comment]
All feedback addressed. Merging
